field/eisenstein/eisenstein_test.go

```diff
@@ -0,0 +1,278 @@
+			b.Mul(a, one)
+			return a.Equal(&b)
+		},
+		genE,
+	))
+
+	properties.Property("add should be commutative", prop.ForAll(
+		func(a, b *ComplexNumber) bool {
+			var c, d ComplexNumber
+			c.Add(a, b)
+			d.Add(b, a)
+			return c.Equal(&d)
+		},
+		genE,
+		genE,
+	))
+
+	properties.Property("add should be assiocative", prop.ForAll(
+		func(a, b, c *ComplexNumber) bool {
+			var d, e ComplexNumber
+			d.Add(a, b).Add(&d, c)
+			e.Add(c, b).Add(&e, a)
+			return e.Equal(&d)
+		},
+		genE,
+		genE,
+		genE,
+	))
+
+	properties.Property("mul should be commutative", prop.ForAll(
+		func(a, b *ComplexNumber) bool {
+			var c, d ComplexNumber
+			c.Mul(a, b)
+			d.Mul(b, a)
+			return c.Equal(&d)
+		},
+		genE,
+		genE,
+	))
+
+	properties.Property("mul should be assiocative", prop.ForAll(
+		func(a, b, c *ComplexNumber) bool {
+			var d, e ComplexNumber
+			d.Mul(a, b).Mul(&d, c)
+			e.Mul(c, b).Mul(&e, a)
+			return e.Equal(&d)
+		},
+		genE,
+		genE,
+		genE,
+	))
+
+	properties.Property("norm should always be positive", prop.ForAll(
+		func(a *ComplexNumber) bool {
+			return a.Norm().Sign() > 0
+		},
+		genE,
+	))
+
+	properties.TestingRun(t, gopter.ConsoleReporter(false))
+}
+
+func TestEisensteinHalfGCD(t *testing.T) {
+
+	t.Parallel()
+	parameters := gopter.DefaultTestParameters()
+	if testing.Short() {
+		parameters.MinSuccessfulTests = nbFuzzShort
+	} else {
+		parameters.MinSuccessfulTests = nbFuzz
+	}
+
+	properties := gopter.NewProperties(parameters)
+
+	genE := GenComplexNumber()
+
+	properties.Property("half-GCD", prop.ForAll(
+		func(a, b *ComplexNumber) bool {
+			res := HalfGCD(a, b)
+			var c, d ComplexNumber
+			c.Mul(b, res[1])
+			d.Mul(a, res[2])
+			d.Add(&c, &d)
+			return d.Equal(res[0])
+		},
+		genE,
+		genE,
+	))
+
+	properties.TestingRun(t, gopter.ConsoleReporter(false))
+}
+
+// GenNumber generates a random integer
+func GenNumber() gopter.Gen {
+	return func(genParams *gopter.GenParameters) *gopter.GenResult {
+		var prime, _ = new(big.Int).SetString("7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffed", 16) // 2^255 - 19
+		elmt, _ := rand.Int(rand.Reader, prime)
+		genResult := gopter.NewGenResult(*elmt, gopter.NoShrinker)
+		return genResult
+	}
+}
+
+// GenComplexNumber generates a random integer
+func GenComplexNumber() gopter.Gen {
+	return gopter.CombineGens(
+		GenNumber(),
+		GenNumber(),
+	).Map(func(values []interface{}) *ComplexNumber {
+		return &ComplexNumber{A0: values[0].(big.Int), A1: values[1].(big.Int)}
+	})
+}
+
+// bench
+func BenchmarkHalfGCD(b *testing.B) {
+	var prime, _ = new(big.Int).SetString("7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffed", 16) // 2^255 - 19
+	a0, _ := rand.Int(rand.Reader, prime)
+	a1, _ := rand.Int(rand.Reader, prime)
+	c0, _ := rand.Int(rand.Reader, prime)
+	c1, _ := rand.Int(rand.Reader, prime)
+	a := ComplexNumber{A0: *a0, A1: *a1}
+	c := ComplexNumber{A0: *c0, A1: *c1}
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		HalfGCD(&a, &c)
```

Comment:
Afaik there was a recommendation to store the result in some variable to ensure that Go compiler doesn't optimize the call out? [See here](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go#:~:text=A%20note%20on%20compiler%20optimisations)